Soft delete collections

pomegranited opened the pull request:
<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

Adds support for soft-deleting and restoring Collections.

As part of this change, we're refactoring the way the Collections-related events are triggered to use Django model signals, so that changes made in Django Admin will be reflected in the Studio search index.

- Which edX user roles will this change impact? Library Authors; not enabled in production by default.

## Supporting information

Related to: https://github.com/openedx/modular-learning/issues/231
Depends on: https://github.com/openedx/openedx-learning/pull/229
Private-ref: [FAL-3809](https://tasks.opencraft.com/browse/FAL-3809)

## Testing instructions

Recommend reviewing the code commit by commit.

1. Reindex your studio content to update any existing Collection documents: `tutor dev run cms ./manage.py cms reindex_studio --experimental`
2. In the Authoring frontend, create a Collection and add some components to it.
3. Open http://meilisearch.local.edly.io:7700/
    Your api key can be found with tutor config printvalue MEILISEARCH_API_KEY
    Search for your collection key and verify that the collection document and its components are returned.
4. Hit the "soft delete" REST API: DELETE http://studio.local.edly.io:8001/api/libraries/v2/<library_key>/collections/<collection_key>/
    Note that the collection no longer appears in the Authoring MFE list of collections.
    Search for your collection key in Meilisearch and verify that no results are found.
5. Hit the "restore" REST API: POST http://studio.local.edly.io:8001/api/libraries/v2/<library_key>/collections/<collection_key>/restore/
   Verify that the collection again appears in the Authoring MFE list of collections.
    Search for your collection key in Meilisearch and verify that the collection document and its components are returned again.
6. Open *Studio* Django Admin http://studio.local.edly.io:8001/admin/oel_collections/collection/
     Disabling a collection should have the same effect as the "soft delete" REST API (step 4).
     Re-enabling a collection should have the same effect as the "restore" REST API (step 5)
7. Delete a Collection via Django Admin.
     This will permanently delete the Collection.
8. Try to hit the "restore" REST API with the deleted collection key: POST http://studio.local.edly.io:8001/api/libraries/v2/<library_key>/collections/<collection_key>/restore/
   This should return a 404.

## Deadline

ASAP

## Author Notes & Concerns

* f760c2b6d561093e163a2d92794578d30b9d8726 changes the Meilisearch ID used for Collections from the int `Collection.pk` to a string generated from the Collection's usage key.
  We needed to do this so that the Collection document could be identified by the information in the `CONTENT_LIBRARY_DELETED` event (`library_key` + `collection_key`), so this document can be removed after the Collection itself is removed.
  Users will need to reindex their studio search content.

pomegranited committed:
refactor: use django signals to trigger LIBRARY_COLLECTION events

openedx-webhooks commented:
Thanks for the pull request, @pomegranited!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

pomegranited commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1103,8 +1100,7 @@ def create_library_collection(
     content_library: ContentLibrary | None = None,
 ) -> Collection:
     """
```
Do we even need `create_library_collection` and `update_library_collection` API methods now? They're just wrappers for the authoring_api methods now that we're not sending events anymore.

pomegranited committed:
temp: use openedx-learning branch

from https://github.com/openedx/openedx-learning/pull/229

pomegranited committed:
refactor: use collection usage_key as search document id

This change standardises the search document "id" to be a meilisearch ID
generated from the usage key, for all types of indexed objects.

This is important for collections so we can locate the collection
document in the search index solely from the data provided by the
LIBRARY_COLLECTION_DELETED event (library_key + collection_key), even if
the collection has been deleted from the database.

pomegranited committed:
refactor: avoid fetching more data than we have to.

* get_library_collection_usage_key and
  searchable_doc_tags_for_collection do not need a Collection object;
  the usage key can be created from the library_key and collection_key.

* updated searchable_doc_for_collection to require the parts of the
  collection usage key + an optional collection. This allows us to
  identify the collection's search document from its usage key without
  requiring an existing Collection object (in case it's been deleted).
  Also removes the edge case for indexing Collections not associated
  with a ContentLibrary -- this won't ever really happen.

pomegranited committed:
feat: remove soft- and hard-deleted collections from search index

pomegranited committed:
feat: adds library_component_usage_key to content_libraries.api

pomegranited committed:
refactor: send CONTENT_OBJECT_ASSOCIATON_CHANGED on django model signals

so that added/removed collections are removed/re-added to component documents.

Special case: When a collection is soft-deleted/restored, we detect this
in the search index and update the collection's component documents
directly, without a CONTENT_OBJECT_ASSOCIATON_CHANGED signal.

pomegranited committed:
feat: adds views to soft-delete and restore a collection

pomegranited commented on requirements/constraints.txt:
```diff
@@ -93,7 +93,7 @@ libsass==0.10.0
 
 # pinning this version to avoid updates while the library is being developed
-openedx-learning==0.11.5
+openedx-learning @ git+https://github.com/open-craft/openedx-learning.git@jill/collection-delete
```
To do: fix this when PR merges.

pomegranited committed:
Merge remote-tracking branch 'origin/master' into jill/collection-delete

ChrisChV commented:
@pomegranited The code looks good.

When testing the "soft delete" and "hard delete" with the admin panel, it don't update the index:

https://www.loom.com/share/115ac6dcc9504200b6c58cf3ed97a358?sid=2280dbda-8be2-4bb2-a630-714e5a2cf5ab

pomegranited committed:
Merge remote-tracking branch 'origin/master' into jill/collection-delete

pomegranited commented:
Hi @ChrisChV 
> When testing the "soft delete" and "hard delete" with the admin panel, it don't update the index:

I found out the issue here -- it has to be done from the Studio Django Admin, not the LMS (PR instructions updated).

I'm not sure how best to deal with it though -- [the libraries app is installed on the LMS](https://github.com/openedx/edx-platform/blob/a2e29596177904d92d91bf9a5927ed295520a443/lms/envs/common.py#L3395-L3399) and the CMS, and the right events get raised no matter whether you use the LMS or Studio Django Admin. However, since the content/search app is only installed in the CMS, so its handlers only get run when the event gets raised in Studio.

If I'm understanding this right, running in production with an event bus would fix this, because the `LIBRARY_COLLECTION_UPDATED` event would be put on the bus for all apps to see.

pomegranited committed:
chore: bumps openedx-learning to 0.13.0

navinkarkera commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -563,20 +577,93 @@ def upsert_library_block_index_doc(usage_key: UsageKey) -> None:
+        # The index or documennt doesn't exist
+        log.exception(err)
+
+    return None
```
`document` is not used. I think you meant to save `index.get_document(id)` in `document` and return it at the end.
```suggestion
    document = None
    try:
        index = client.get_index(STUDIO_INDEX_NAME)
        document = index.get_document(id)
    except (MeilisearchError, MeilisearchApiError) as err:
        # The index or documennt doesn't exist
        log.exception(err)

    return document
```

pomegranited replied:
Oh wow, thank you for spotting those typos @navinkarkera ! Fixed with e92f40544e.

navinkarkera reviewed (approved):
@pomegranited Nice work! :100:  Left a small nit.

- [x] I tested this: Soft-deleted collections and restored it, verified events.
- [x] I read through the code
- [x] ~I checked for accessibility issues~
- [x] Includes documentation

ChrisChV reviewed (approved):
@pomegranited Looks good :+1:
- [X] I tested this: I followed the testing instructions
- [X] I read through the code and considered the security, stability and performance implications of the changes.
- [X] Includes tests for bugfixes and/or features added.
- [x] Includes documentation
- [X] Includes fixtures that create objects required for manual testing.

pomegranited committed:
fix: fix _get_document_from_index

typos in code and comments

ChrisChV committed:
Merge branch 'master' into jill/collection-delete

ChrisChV committed:
Merge branch 'master' into jill/collection-delete

ChrisChV committed:
Merge branch 'master' into jill/collection-delete

ChrisChV committed:
Merge branch 'master' into jill/collection-delete
